packages/fuels-macros/src/parse_utils.rs

```diff
@@ -80,3 +81,118 @@ where
+            GenericParam::Lifetime(lifetime) => {
+                Err(Error::new_spanned(lifetime, "Lifetimes not supported"))
+            }
+            GenericParam::Const(const_generic) => Err(Error::new_spanned(
+                const_generic,
+                "Const generics not supported",
+            )),
+        })
+        .collect()
+}
+
+pub(crate) struct Members {
+    names: Vec<Ident>,
+    types: Vec<TokenStream>,
+    fuels_types_path: TokenStream,
+}
+
+impl Members {
+    pub(crate) fn names(&self) -> impl Iterator<Item = &Ident> + '_ {
+        self.names.iter()
+    }
+
+    pub(crate) fn names_as_strings(&self) -> impl Iterator<Item = TokenStream> + '_ {
+        self.names().map(|ident| {
+            let name = ident.to_string();
+            quote! {#name.to_string()}
+        })
+    }
+    pub(crate) fn param_type_calls(&self) -> impl Iterator<Item = TokenStream> + '_ {
+        let fuels_types_path = self.fuels_types_path.to_token_stream();
+        self.types.iter().map(move |ty| {
+            quote! { <#ty as #fuels_types_path::traits::Parameterize>::param_type() }
+        })
+    }
+}
+
+pub(crate) fn extract_struct_members(
+    fields: DataStruct,
+    fuels_types_path: TokenStream,
+) -> syn::Result<Members> {
+    let named_fields = match fields.fields {
+        Fields::Named(named_fields) => Ok(named_fields.named),
+        Fields::Unnamed(fields) => Err(Error::new_spanned(
+            fields.unnamed,
+            "Tuple-like structs not supported",
+        )),
+        _ => {
+            panic!("This cannot happen in valid Rust code. Fields::Unit only appears in enums")
+        }
+    }?;
+
+    let (names, types) = named_fields
+        .into_iter()
+        .map(|field| {
+            let name = field
+                .ident
+                .expect("FieldsNamed to only contain named fields!.");
+            let ty = field.ty.into_token_stream();
+            (name, ty)
+        })
+        .unzip();
+
+    Ok(Members {
+        names,
+        types,
+        fuels_types_path,
+    })
+}
+
+pub(crate) fn extract_enum_members(
+    data: DataEnum,
+    fuels_types_path: TokenStream,
+) -> syn::Result<Members> {
+    let components = data.variants.into_iter().map(|variant: Variant| {
+        let name = variant.ident;
+
+        let ttype = match variant.fields {
+            Fields::Unnamed(fields_unnamed) => {
+                if fields_unnamed.unnamed.len() != 1 {
+                    return Err(Error::new(
+                        fields_unnamed.paren_token.span,
+                        "Must have exactly one element",
+                    ));
+                }
+                fields_unnamed.unnamed.into_iter().next()
+            }
+            Fields::Unit => None,
+            Fields::Named(named_fields) => {
+                return Err(Error::new_spanned(
+                    named_fields,
+                    "Struct-like enum variants are not supported!",
```

Comment:
```suggestion
                    "Struct-like enum variants are not supported.",
```

Incredibly minor: let's try and consistently stick with period (`.`) instead of exclamation marks (`!`) for error messages 😁